torch/_inductor/autoheuristic.py

```diff
@@ -0,0 +1,154 @@
+import json
+import os
+
+from typing import Any, Callable, List, Tuple, Union
+
+import torch
+
+from torch._inductor.runtime.runtime_utils import cache_dir
+
+Feedback = Union[int, float]
+Choice = str
+
+
+class _Feedback:
+    def __init__(self, feedback_fn: Callable[[Choice], Feedback]) -> None:
+        self.feedback_fn = feedback_fn
+
+    def __call__(self, choice: Choice) -> Feedback:
+        return self.feedback_fn(choice)
+
+
+class LocalFeedback(_Feedback):
+    def __init__(self, feedback_fn: Callable[[Choice], Feedback]) -> None:
+        super().__init__(feedback_fn)
+
+
+class GlobalFeedback(_Feedback):
+    # TODO: will be supported later
+    def __init__(self, feedback_fn: Callable[[Choice], Feedback]) -> None:
+        super().__init__(feedback_fn)
+
+
+class AHFeature:
+    def __init__(
+        self, name: str, value: Union[int, float, str], is_categorical: bool = False
+    ) -> None:
+        self.name = name
+        self.value = value
+        self.is_categorical = is_categorical
+
+
+class AHContext:
+    features: List[AHFeature]
+
+    def __init__(self) -> None:
+        self.features = []
+
+    def add_feature(
+        self, name: str, value: Union[int, float, str], is_categorical: bool = False
+    ) -> None:
+        self.features.append(AHFeature(name, value, is_categorical=is_categorical))
+
+
+class InconsistentMetadata(Exception):
+    pass
+
+
+class AutoHeuristic:
+    def __init__(
+        self,
+        fallback: Choice,
+        choices: List[Choice],
+        feedback: Union[LocalFeedback, GlobalFeedback],
+        context: AHContext,
+        name: str,
+    ) -> None:
+        self.fallback = fallback
+        self.choices = choices
+        self.feedback = feedback
+        self.context = context
+        self.name = name
+        self.features = context.features
+
+        if torch._inductor.config.autoheuristic_mode == "COLLECT_DATA" and isinstance(
```

Comment:
You don't need to implement this now, but we'll want the ability to specify which heuristic/heuristics we're collecting data for eventually. So, for example, perhaps eventually it might be something like `AUTOHEURISTIC_MODE="collect:pad_mm,foo,bar"`

Reply:
I agree. I made a note to myself and will implement this once I try autoheuristic on something that is not pad_mm.

Reply:
Folks often will also leave a `# todo(alnis)` in the code.